tls/s2n_server_key_exchange.c

```diff
@@ -29,71 +29,70 @@
 #include "utils/s2n_safety.h"
 #include "utils/s2n_random.h"
 
-static int s2n_ecdhe_server_key_recv(struct s2n_connection *conn);
-static int s2n_dhe_server_key_recv(struct s2n_connection *conn);
-static int s2n_ecdhe_server_key_send(struct s2n_connection *conn);
-static int s2n_dhe_server_key_send(struct s2n_connection *conn);
 static int s2n_write_signature_blob(struct s2n_stuffer *out, const struct s2n_pkey *priv_key, struct s2n_hash_state *digest);
 
 int s2n_server_key_recv(struct s2n_connection *conn)
 {
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_ECC) {
-        GUARD(s2n_ecdhe_server_key_recv(conn));
-    } else {
-        GUARD(s2n_dhe_server_key_recv(conn));
-    }
-    return 0;
-}
-
-static int s2n_ecdhe_server_key_recv(struct s2n_connection *conn)
-{
+    struct s2n_hash_state *signature_hash = &conn->secure.signature_hash;
+    const struct s2n_key_exchange_algorithm *key_exchange = conn->secure.cipher_suite->key_exchange_alg;
     struct s2n_stuffer *in = &conn->handshake.io;
-    struct s2n_blob ecdhparams = {0};
-    struct s2n_blob signature = {0};
-    uint16_t signature_length;
+    struct s2n_blob data_to_sign = {0};
 
-    /* Read server ECDH params and calculate their hash */
-    GUARD(s2n_ecc_read_ecc_params(&conn->secure.server_ecc_params, in, &ecdhparams));
+    /* Read and process the KEX data */
+    notnull_check(key_exchange->server_key_recv);
+    GUARD(key_exchange->server_key_recv(conn, &data_to_sign));
 
+    /* Add common signature data */
     if (conn->actual_protocol_version == S2N_TLS12) {
         s2n_hash_algorithm hash_algorithm;
         s2n_signature_algorithm signature_algorithm;
         GUARD(s2n_get_signature_hash_pair_if_supported(in, &hash_algorithm, &signature_algorithm));
-
-        GUARD(s2n_hash_init(&conn->secure.signature_hash, hash_algorithm));
+        GUARD(s2n_hash_init(signature_hash, hash_algorithm));
     } else {
-        GUARD(s2n_hash_init(&conn->secure.signature_hash, conn->secure.conn_hash_alg));
+        GUARD(s2n_hash_init(signature_hash, conn->secure.conn_hash_alg));
     }
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, ecdhparams.data, ecdhparams.size));
+    GUARD(s2n_hash_update(signature_hash, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
+    GUARD(s2n_hash_update(signature_hash, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
+
+    /* Add KEX specific data */
+    GUARD(s2n_hash_update(signature_hash, data_to_sign.data, data_to_sign.size));
 
     /* Verify the signature */
+    struct s2n_blob signature = {0};
```

Comment:
Minor nitpick here: if you declared this slightly later, you could init it directly with the computed values
```
 struct s2n_blob signature = {.size = signature_length, .data = s2n_stuffer_raw_read(in, signature.size)};